module/modmanager/manager.go

```diff
@@ -0,0 +1,300 @@
+						return mgr.AddResource(ctx, config.ServiceConfigToShared(cfg), DepsToNames(deps))
+					},
+				})
+			}
+		default:
+			mgr.logger.Errorf("invalid module type: %s", api.Subtype.Type)
+		}
+	}
+
+	return nil
+}
+
+// AddResource tells a component module to configure a new component.
+func (mgr *Manager) AddResource(ctx context.Context, cfg config.Component, deps []string) (interface{}, error) {
+	mgr.mu.Lock()
+	defer mgr.mu.Unlock()
+	module := mgr.getModule(cfg)
+	if module == nil {
+		return nil, errors.Errorf("no module registered to serve resource api %s and model %s", cfg.ResourceName().Subtype, cfg.Model)
+	}
+
+	cfgProto, err := config.ComponentConfigToProto(&cfg)
+	if err != nil {
+		return nil, err
+	}
+	_, err = module.client.AddResource(ctx, &pb.AddResourceRequest{Config: cfgProto, Dependencies: deps})
+	if err != nil {
+		return nil, err
+	}
+	mgr.rMap[cfg.ResourceName()] = module
+
+	c := registry.ResourceSubtypeLookup(cfg.ResourceName().Subtype)
+	if c == nil || c.RPCClient == nil {
+		mgr.logger.Warnf("no built-in grpc client for modular resource %s", cfg.ResourceName())
+		return rdkgrpc.NewForeignResource(cfg.ResourceName(), module.conn), nil
+	}
+	nameR := cfg.ResourceName().ShortName()
+	return c.RPCClient(ctx, module.conn, nameR, mgr.logger), nil
+}
+
+// ReconfigureResource updates/reconfigures a modular component with a new configuration.
+func (mgr *Manager) ReconfigureResource(ctx context.Context, cfg config.Component, deps []string) error {
+	mgr.mu.RLock()
+	defer mgr.mu.RUnlock()
+	module := mgr.getModule(cfg)
+	if module == nil {
+		return errors.Errorf("no module registered to serve resource api %s and model %s", cfg.ResourceName().Subtype, cfg.Model)
+	}
+
+	cfgProto, err := config.ComponentConfigToProto(&cfg)
+	if err != nil {
+		return err
+	}
+	_, err = module.client.ReconfigureResource(ctx, &pb.ReconfigureResourceRequest{Config: cfgProto, Dependencies: deps})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// NeedsModule returns true if a component/service config WOULD be handled by a module.
+func (mgr *Manager) NeedsModule(cfg config.Component) bool {
+	mgr.mu.RLock()
+	defer mgr.mu.RUnlock()
+	return mgr.getModule(cfg) != nil
+}
+
+// IsModularResource returns true if an existing resource IS handled by a module.
+func (mgr *Manager) IsModularResource(name resource.Name) bool {
+	mgr.mu.RLock()
+	defer mgr.mu.RUnlock()
+	_, ok := mgr.rMap[name]
+	return ok
+}
+
+// RemoveResource requests the removal of a resource from a module.
+func (mgr *Manager) RemoveResource(ctx context.Context, name resource.Name) error {
+	mgr.mu.Lock()
+	defer mgr.mu.Unlock()
+	module, ok := mgr.rMap[name]
+	if !ok {
+		return errors.Errorf("resource %+v not found in module", name)
+	}
+
+	_, err := module.client.RemoveResource(ctx, &pb.RemoveResourceRequest{Name: name.String()})
+	return err
+}
+
+func (mgr *Manager) getModule(cfg config.Component) *module {
```

Comment:
[minor] do not rely on nilness as a means of not found. use an explicit boolean param